sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/multi/MultiServiceManager.java

```diff
@@ -0,0 +1,282 @@
+    // When a client is added, if we're already registered then invoke 'registered()' manually against that client
+    private boolean isRegistered;
+
+    public MultiServiceManager() {
+        this.isRegistered = false;
+    }
+
+    /**
+     * Returns the original names of currently available services.
+     */
+    public Collection<String> getServiceNames() {
+        Collection<String> serviceNames = new TreeSet<>(); // Alphabetical order
+        rlock.lock();
+        try {
+            serviceNames.addAll(services.keySet());
+        } finally {
+            rlock.unlock();
+        }
+        return serviceNames;
+    }
+
+    /**
+     * Returns the specified service by its original name, or an empty {@code Optional} if it's not found.
+     */
+    public Optional<AbstractScheduler> getService(String serviceName) {
+        rlock.lock();
+        try {
+            return Optional.ofNullable(services.get(serviceName));
+        } finally {
+            rlock.unlock();
+        }
+    }
+
+    /**
+     * Returns the specified service by its sanitized name, or an empty {@code Optional} if it's not found.
+     *
+     * A sanitized service name has had its slashes removed. For example,
+     * "{@code /path/to/service}" => "{@code path.to.service}". Sanitized names may be used in e.g. URLs that contain
+     * the service name.
+     */
+    public Optional<AbstractScheduler> getServiceSanitized(String sanitizedServiceName) {
+        rlock.lock();
+        try {
+            String originalName = sanitizedServiceNames.get(sanitizedServiceName);
+            if (originalName == null) {
+                // Unknown sanitized name
+                return Optional.empty();
+            }
+            return Optional.ofNullable(services.get(originalName));
+        } finally {
+            rlock.unlock();
+        }
+    }
+
+    /**
+     * Adds or replaces a running service according to the name in its ServiceSpec. Note: If the service was marked for
+     * uninstall via {@link #uninstallService(String)}, it should continue to be added across scheduler restarts in
+     * order for uninstall to complete. It should only be omitted after the uninstall callback has been invoked for it.
+     *
+     * @param service the client to add, or to replace an existing client with if their service names match
+     * @return {@code this}
+     * @throws IllegalArgumentException if the service name collides with an existing service after reducing slashes to
+     *                                  periods
+     */
+    public MultiServiceManager putService(AbstractScheduler service) {
+        String originalName = service.getServiceSpec().getName();
+        String sanitizedName = CommonIdUtils.toSanitizedServiceName(originalName);
+        rwlock.lock();
+        try {
+            // NOTE: If the service is uninstalling, it should already be passed to us as an UninstallScheduler.
+            // See SchedulerBuilder.
```

Comment:
I guess this occurs through some factory type thing that detects whether uninstall is still active for the service?

Reply:
Yes, `SchedulerBuilder` checks for a few cases (between service removal vs global uninstall) here: https://github.com/mesosphere/dcos-commons/blob/master/sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/SchedulerBuilder.java#L281-L314